### PR TITLE
Support Deployments and StatefulSet

### LINKED ISSUE #49: Support StatefulSet only For Historicals,MM
### Description
- Currently the operator provisions statefulset for all the druid processes, regardless of their nature (stateless/statefull )
- Brokers, routers, coordinators, overlords are stateless in nature and don't require a persistence. 
- Statefulsets have their own challenges such as sts does not support rollbacks ( since their is no replicaset associated with sts ), pvc's remain in not reference/hanging state during scale up/down of pods ( though ideally operator should address this problem https://github.com/druid-io/druid-operator/issues/43 ). At time sts need to be manually deleted using cascade options.
- The operator should support deployment objects for all process accept MM/Historicals.
- The current druid_controller can reconcile against deployment events, we don't need a separate controller.
- As discussed earlier https://github.com/druid-io/druid-operator/issues/17, mostly logging solutions such as fluentd, dd agents running as daemonsets can push logging to external datastores for querying. Using EBS for logging can be an option not necessity, 
- https://github.com/druid-io/druid-operator/blob/master/pkg/controller/druid/handler.go#L537 this func needs to be split up for seperate podSpec and podTemplateSpec which can be used as common to both deployments.sts. 
- https://github.com/druid-io/druid-operator/blob/master/pkg/controller/druid/handler.go#L133, a conditional check on the basis of nodeType can be used whether to deploy deployment or sts.
- Implementing these changes on existing cluster may require manual intervention. 
------------------------------------------------------------------------------------------------------------------------------
@himanshug

### PR DESCRIPTION
**Pull request by AdheipSingh**
Fixes #49 .

### Description

- As discussed in the issue #49, this PR shall make sure operator supports both deployments as well as statefulsets, giving users more flexibility. 

- In the ```NodeSpec``` have added a key ```kind``` which can take inputs for deployments or statefulsets.
- This key is a ```Required``` param and is validated by the operator.
- The ```makeStatefulSet``` func has been split up so that the underlying ```podSpec``` remains common. Hence less code to maintain, plus in case we add more features to the podSpec it remains common to both deployments and statefulsets.
- Status has been updated to update deployments.
- Update is same as earlier, incase of rollingDeploy true, we have incremental updates.
<hr>

This PR has:
- [x ] been tested on a real K8S cluster to ensure creation of a brand new Druid cluster works.
- [x ] been tested for backward compatibility on a real K*S cluster by applying the changes introduced here on an existing Druid cluster. If there are any backward incompatible changes then they have been noted in the PR description.
- [x ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added documentation for new or modified features or behaviors.

<hr>

##### Key changed/added files in this PR
 * `types.go`
 * `handler.go`
 * `tiny-cluster.yaml`

### CODE REVIEW

File: pkg/apis/druid/v1alpha1/druid_types.go [hunk 1]
```diff
@@ -141,6 +141,9 @@ type DruidNodeSpec struct {
 	// Required: Port used by Druid Process
 	DruidPort int32 `json:"druid.port"`
 
+	// Required: Define kind for node. Deployment or Statefulsets are supported
+	Kind string `json:"kind"`
```
Comment (himanshug) [inline pkg/apis/druid/v1alpha1/druid_types.go:145]: can we make it optional and default to `StatefulSet` to keep things backward compatible and also to not necessarily having to specify this.

Comment (AdheipSingh) [reply pkg/apis/druid/v1alpha1/druid_types.go:145]: sure, makes sense to have backward compatible. Ill do the changes :)

**Comment by AdheipSingh** [comment]
@himanshug any updates ?

**Comment by himanshug** [comment]
@AdheipSingh sorry, I haven't had a chance to look at it till now, will review it this week for sure. I am hoping to get some more time to spend on druid-operator in future :)

**Comment by AdheipSingh** [comment]
> @AdheipSingh sorry, I haven't had a chance to look at it till now, will review it this week for sure. I am hoping to get some more time to spend on druid-operator in future :)

no hurries, just wanted to check :)
Thanks

File: pkg/apis/druid/v1alpha1/druid_types.go [hunk 2]
```diff
@@ -141,6 +141,10 @@ type DruidNodeSpec struct {
 	// Required: Port used by Druid Process
 	DruidPort int32 `json:"druid.port"`
 
+	// Optional: Define kind for node. Deployment or Statefulsets are supported,
+	// Defaults to statefulsets.
```
Comment (himanshug) [inline pkg/apis/druid/v1alpha1/druid_types.go:145]:
```suggestion
	// Defaults to statefulsets.
	// Note that, volumeClaimTemplates are ignored when kind=Deployment
```

File: pkg/controller/druid/handler.go [hunk 3]
```diff
@@ -408,6 +446,25 @@ func isStsFullyDeployed(sdk client.Client, name string, drd *v1alpha1.Druid) (bo
...
+// Checks if all replicas corresponding to latest updated sts have been deployed
+func isDeployFullyDeployed(sdk client.Client, name string, drd *v1alpha1.Druid) (bool, error) {
+	deploy := makeDeploymentEmptyObj()
+	if err := sdk.Get(context.TODO(), *namespacedName(name, drd.Namespace), deploy); err != nil {
+		e := fmt.Errorf("failed to get [Deployment:%s] due to [%s]", name, err.Error())
+		logger.Error(e, e.Error(), "name", drd.Name, "namespace", drd.Namespace)
+		sendEvent(sdk, drd, v1.EventTypeWarning, "GET_FAIL", e.Error())
+		return false, e
+	} else {
+		if deploy.Status.ReadyReplicas != deploy.Status.Replicas {
```
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply pkg/controller/druid/handler.go:456]: yes, i have been running this in my cluster, this is checking if replicas desired have come back in ready state. Behaves same as sts during upgrades.

File: pkg/controller/druid/handler.go [hunk 4]
```diff
@@ -131,25 +133,47 @@ func deployDruidCluster(sdk client.Client, m *v1alpha1.Druid) error {
...
-			}
+				if deployCreateUpdateStatus == resourceUpdated {
+					return nil
+				}
 
-			// Check StatefulSet rolling update status, if in-progress then stop here
-			done, err := isStsFullyDeployed(sdk, nodeSpecUniqueStr, m)
-			if !done {
+				// Check Deployment rolling update status, if in-progress then stop here
+				done, err := isDeployFullyDeployed(sdk, nodeSpecUniqueStr, m)
```
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply pkg/controller/druid/handler.go]: done

**Pull request merged**